internal/controllers/vap_controller.go

```diff
@@ -0,0 +1,180 @@
+
+func (v *vapController) Reconcile(ctx context.Context, req ctrl.Request) (reconcile.Result, error) {
+	v.log.Info("Starting ValidatingAdmissionPolicy reconciliation...", "request", req.String())
+
+	err := v.reconcileVAP(ctx)
+
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, err
+}
+
+func (v *vapController) reconcileVAP(ctx context.Context) error {
+	vap := NewVMDeleteProtectionValidatingAdmissionPolicy()
+	foundVAP := &admissionregistrationv1.ValidatingAdmissionPolicy{}
+
+	objKey := client.ObjectKey{Name: vap.Name}
+	err := v.client.Get(ctx, objKey, foundVAP)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			err = v.client.Create(ctx, vap)
+			if err != nil {
+				return fmt.Errorf("failed to create VAP %v", err)
+			}
+			return err
+		}
+	}
+
+	if requiresUpdate(vap.Labels, foundVAP.Labels) {
+		foundVAP.Labels = vap.Labels
+		err = v.client.Update(ctx, foundVAP)
+		if err != nil {
+			return err
+		}
+	}
+
+	if requiresUpdate(foundVAP.Spec, vap.Spec) {
+		foundVAP.Spec = *vap.Spec.DeepCopy()
+		err = v.client.Update(ctx, foundVAP)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func requiresUpdate[T any](old, new T) bool {
+	if !reflect.DeepEqual(old, new) {
+		return true
+	}
+	return false
+}
+
+func NewVMDeleteProtectionValidatingAdmissionPolicy() *admissionregistrationv1.ValidatingAdmissionPolicy {
+	failPolicy := admissionregistrationv1.Fail
+
+	return &admissionregistrationv1.ValidatingAdmissionPolicy{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "ValidatingAdmissionPolicy",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: virtualMachineDeleteProtectionPolicyName,
+			Labels: map[string]string{
+				common.AppKubernetesManagedByLabel: common.AppKubernetesManagedByValue,
+				common.AppKubernetesVersionLabel:   env.GetOperatorVersion(),
+				common.AppKubernetesComponentLabel: vapControllerName,
+			},
+		},
+		Spec: admissionregistrationv1.ValidatingAdmissionPolicySpec{
+			FailurePolicy: &failPolicy,
+			MatchConstraints: &admissionregistrationv1.MatchResources{
+				ResourceRules: []admissionregistrationv1.NamedRuleWithOperations{
+					{
+						RuleWithOperations: admissionregistrationv1.RuleWithOperations{
+							Operations: []admissionregistrationv1.OperationType{
+								admissionregistrationv1.Delete,
+							},
+							Rule: admissionregistrationv1.Rule{
+								APIGroups:   []string{"kubevirt.io"},
+								APIVersions: []string{"*"},
+								Resources:   []string{"virtualmachines"},
+							},
+						},
+					},
+				},
+			},
+			Variables: []admissionregistrationv1.Variable{
+				{
+					Name:       "label",
+					Expression: `string('kubevirt.io/vm-delete-protection')`,
+				},
+			},
+			Validations: []admissionregistrationv1.Validation{
+				{
+					Expression:        `(!has(oldObject.metadata.labels) || !(variables.label in oldObject.metadata.labels) || !oldObject.metadata.labels[variables.label].matches('^(true|True)$'))`,
```

Comment:
Would it be enough to test that the label exists, instead of checking its value?

Reply:
I've considered clearer from the user point of view. My idea is: by forcing the user to add the "True" value, we make sure that user really wants the protection enabled, and I think it will avoid confusions. That being said, I'm not against it if we find it out better.